### ISSUE TITLE
Xcode 14 beta: Thread running at QOS_CLASS_USER_INTERACTIVE waiting on a lower QoS thread running at QOS_CLASS_DEFAULT. Investigate ways to avoid priority inversions

### ISSUE THREAD

**Issue by groue** [groue/GRDB.swift#1234]
> Thread running at QOS_CLASS_USER_INTERACTIVE waiting on a lower QoS thread running at QOS_CLASS_DEFAULT. Investigate ways to avoid priority inversions

This message can appear as a runtime warning (purple ⚠️ sign), in Xcode 14 beta, inside the internal `GRDB.Pool.get()` method:

```swift
final class Pool<T> {
    func get() throws -> (element: T, release: () -> Void) {
        try barrierQueue.sync {
            itemsSemaphore.wait() // <- ⚠️ Thread running at QOS_CLASS_USER_INTERACTIVE…
```

It happens when the application performs a synchronous database read, from the main thread, with `DatabasePool`.

---

**To remove this warning, instruct the database to run with higher priority:**

```swift
var configuration = Configuration()
// As we plan to access the database from the main thread, make
// sure it runs with an appropriate quality of service:
configuration.qos = .userInitiated
let dbPool = try DatabasePool(path: ..., configuration: configuration)
```

---

This warning is a new tool in Xcode 14 beta, the **Thread Performance Checker**. You enable it in Xcode > Edit Scheme… > Run > Diagnostics > Runtime API checking > check "Thread Performance Checker".

Thank you Apple for warning us that we misuse our threads :-)

---

We'll probably raise the default quality of service of GRDB databases in a future release. If we do not, many Xcode 14 users will start to see this runtime warning, and I don't want to ship a library that warns on first use:

- I do not want to deal with the support requests this would create.
- SQLite is very fast, simple apps ought to remain simple, and nobody should be shamed for accessing the db from the main thread.

Please feel free to comment below, if you think we can be more subtle than raising all databases to `.userInitiated` 😅

**Comment by groue** [comment]
Fixed by #1235.

**Comment by groue** [comment]
The fix has shipped in v5.25.0

**Comment by peterhoucll** [comment]
Hi, I upgraded GRDB to 4.1.0 and the problem is not solved in Xcode 14.

**Comment by groue** [comment]
Hello @peterhoucll. Please provide a reproducing sample code, if you have time. This would greatly help.

**Comment by peterhoucll** [comment]
Hi, after checking, the warning will appear only in the simulator, thanks.

**Comment by groue** [comment]
All right @peterhoucll. This remains an interesting warning, even if it only happens in the simulator. Again, only if you have the time, and you can see that the warning happens from within GRDB-related code, please share.

**Comment by yeahphil** [comment]
Hello @groue, I'm still able to repro this in the simulator AND on devices using GRDB 6.0.0. Here's a brief description of where I'm seeing the warning. I haven't tried to put together a sample project that reproduces the issue, but let me know if that would be helpful and I'd be happy to do so.

We see this occur when using DatabasePool and calling `read()`. The warning comes from `Pool.get()`, at this `DispatchSemaphore.wait()` call:

```
    func get() throws -> (element: T, release: () -> Void) {
        try barrierQueue.sync {
            itemsSemaphore.wait() // <--- "Thread Performance Checker: Thread running at QOS_CLASS_USER_INTERACTIVE waiting on a lower QoS thread running at QOS_CLASS_DEFAULT."
            itemsGroup.enter()
            // ...method continues
```

That's happening for a tiny minority of reads, and I'm sure there other conditions necessary for it to warn. Again if it would be helpful, I'd be happy to investigate further.

**Comment by groue** [comment]
Hello @yeahphil! I'm not investigating the warning myself, because it's been a long time I could not see it. I simply don't have the opportunity to put in under scrutiny!

Assuming you don't specify a custom quality of service (QoS) or target DispatchQueue in the `configuration` of your DatabasePool, all database operations *should* run at `.userInitiated` (the default `Configuration.qos`).

This QoS does not trigger any runtime warning when it interacts with `.userInteractive`.

Here, the warning says "Thread running at QOS_CLASS_USER_INTERACTIVE waiting on a lower QoS thread running at QOS_CLASS_DEFAULT". So *something* in GRDB is running at QoS `.default` instead of the expected QoS `.userInitiated`. This looks like a GRDB bug - but until we understand how the warning happens, it's hard to say!

**Comment by yeahphil** [comment]
OK, I'll see if I can put together a better informed report and update. It might be a few days before I can get back to it.

We're not specifying a custom QoS or target queue when setting up the DatabasePool. There ARE a couple places where the app uses a DispatchQueue with .default QoS... but it doesn't look like the warning is coming from anything that would be scheduled on those.

I wonder if this could happen if we're doing other work concurrently, exhaust the reader pool, and have to wait. Does that seem like a possibility to you?

**Comment by groue** [comment]
If I try to understand:

- Application calls `DatabasePool.read` (probably from some `.userInteractive` queue)
- which calls `Pool.get`
- which calls `barrierQueue.sync` (this queue has `.userInitiated` qos, and the `sync` method much probably reuses the current thread anyway)
- which calls `itemsSemaphore.wait()`

It's a short stack trace. Everything is synchronous.

So, `itemsSemaphore.wait()` blocks, and we have a runtime warning "waiting on a lower QoS thread running at QOS_CLASS_DEFAULT".

Which is this lower QoS thread? Since the semaphore is blocking, then five other threads are reading. Much probably, one of them, or all of them, run at qos `.default` (I don't know what triggers the warning exactly). Let's assume one is enough.

So the `DatabasePool` is reading from some low QoS thread.

It can not be an asynchronous read, because all asynchronous reads call `Pool.asyncGet` which spawns async jobs with `.userInitiated` qos.

Conclusion: some low QoS thread is performing a synchronous read.

- Either this low QoS read is started by some internal GRDB stuff. And this is a GRDB bug, because the library is supposed to honor the pool configuration, and use the `.userInitiated` qos.
- Either this low QoS read is started by the application. And now, who is responsible for the warning? Isn't it the application itself 😈 ?

**Comment by yeahphil** [comment]
Ah, that makes sense now -- of course you wouldn't expect the warning to be triggered on calls where the `itemsSemaphore.wait()` call doesn't actually block.

Thanks for the detailed followup. It sounds like this is very likely an issue in the application code and not GRDB. I'll report back if I find anything GRDB-related while cleaning this up.

By the way, I realize now that I misunderstood the warning. It's easy to go through the stack trace from the thread checker and verify that it's all running at `.userInitiated`/`.userInteractive` which had me confused. But if I now understand correctly, what the warning means is, here's a stack for the higher priority queue that's being blocked. Somewhere else, there's a pair of `wait()` call executing on a queue at `.default` priority that caused it to block, but the stack trace for that call is (not surprisingly) not available.

**Comment by groue** [comment]
Yes, I'm glad it starts to make sense as well. Thinking further, I think that the condition that triggers the warning is:

- `DispatchSemaphore.wait()` is blocking on a high QoS thread.
- `DispatchSemaphore.signal()` is called on a low QoS thread.

I mean that if I had to implement the warning, I may have chosen this technique. I wonder if there exists online documentation about those warnings, BTW 😅 

So I focused on code paths that lead to `signal()`. It's always on the same thread as `wait()`, so it's always `.userInitiated` for DatabasePool asynchronous read methods, and whatever QoS is used by the app or GRDB when it performs a synchronous read.

I don't know if it's possible to ask Xcode to breakpoint on the warning. If so, we may be able to see the faulty `signal()` that happens on a low QoS thread? 🤩

**Comment by groue** [comment]
> I don't know if it's possible to ask Xcode to breakpoint on the warning.

It looks like it is possible there, in the breakpoints tab:
<img width="239" alt="Capture d’écran 2022-10-17 à 08 30 56" src="https://user-images.githubusercontent.com/54219/196104749-81e817ae-42e0-46b4-a496-b3c77c476c21.png">

**Comment by groue** [comment]
> I added this line `configuration.qos = .userInitiated`

This is already [the default](https://github.com/groue/GRDB.swift/blob/c3cc998fc14efa356259e5d01d8910e82de080ad/GRDB/Core/Configuration.swift#L316) on GRDB 6.6.0.

> it fixed my issue on real device but on simulator I still get it !

I don't what to tell you. I already gave much information and advices in this issue, and in #1310 as well, but it does not seem like you take care of using them in any way. None of your answers show the slightest sign that you take care of answers. Maybe read again previous messages, follow advice, set breakpoints, inspect how your app is creating threads, and you'll solve your issue.

**Comment by SF-Simon** [comment]
I am also troubled by this issue. After troubleshooting, I found that my query statement is running in the _background_ thread, which is a background task.

In the end, I used `DispatchQueue.global(qos: .userInitiated){}` to execute the query code, and there were no more thread warnings.

So if anyone else encounters it, it is recommended to check the query statement. This is something that was not mentioned in the previous discussion.
